src/btree/bt_delete.c

```diff
@@ -293,19 +293,51 @@ __wt_delete_page_rollback(WT_SESSION_IMPL *session, WT_REF *ref)
  *     IDs in the delete info. Since we're called at the end of recovery there's no need to lock the
  *     ref or worry about races.
  */
-static void
+static int
 __delete_redo_window_cleanup_internal(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_REF *child;
+    uint64_t sleep_usecs, yield_count;
+    bool busy;
 
+    /* This function should be called during recovery only. */
+    WT_ASSERT(session, F_ISSET(S2C(session), WT_CONN_RECOVERING));
     WT_ASSERT(session, F_ISSET(ref, WT_REF_FLAG_INTERNAL));
     if (ref->page != NULL) {
         WT_INTL_FOREACH_BEGIN (session, ref->page, child) {
-            if (child->state == WT_REF_DELETED && child->page_del != NULL)
+            /*
+             * Cleanup any page delete structures in the child. At this stage, it can contain stale
+             * data we should not rely on.
+             */
+            if (child->page_del != NULL) {
                 __cell_redo_page_del_cleanup(session, ref->page->dsk, child->page_del);
+                /*
+                 * Since the page might be in memory during RTS, the window cleanup may not have any
+                 * work. In order to avoid reading incorrect values later on, move the page back to
+                 * disk by evicting it. When it will be read again, the window cleanup should clear
+                 * any stale data.
+                 */
+                if (child->state == WT_REF_MEM) {
+                    for (sleep_usecs = yield_count = 0;;) {
+                        WT_RET(__wt_hazard_set(session, child, &busy));
+                        if (!busy) {
+                            /* The page is expected to be clean. */
+                            WT_ASSERT(session, !__wt_page_is_modified(ref->page));
+                            WT_RET_BUSY_OK(__wt_page_release_evict(session, child, 0));
```

Comment:
The `__wt_page_release_evict` can fail with EBUSY if any other thread is operating on it. At this stage, other than statlog-server none of the sessions should operate on it. How about retrying on a page that failed to evict?

```
ret = __wt_page_release_evict(session, child, 0);
if (ret == 0 || ret != EBUSY)
    break;
```

Reply:
I think the code should be a bit different:
```
-                            WT_RET_BUSY_OK(__wt_page_release_evict(session, child, 0));
-                            break;
+                            ret = __wt_page_release_evict(session, child, 0);
+                            if(ret == 0)
+                                break;
+                            if(ret != EBUSY)
+                                return (ret);
```
If `ret` is `0`, we can break from the rety loop and move on to the next page.
If `ret` is `EBUSY`, keep trying, don't exit the loop.
If `ret` is different from `0` and `EBUSY`, it's another error, we should quit.

Reply:
It could probably be improved this way:
```
-                            WT_RET_BUSY_OK(__wt_page_release_evict(session, child, 0));
-                            break;
+                            ret = __wt_page_release_evict(session, child, 0);
+                            WT_RET_BUSY_OK(ret);
+                            if(ret == 0)
+                                break;
```

Reply:
Thanks for the changes. Yes, something, as proposed, will be good to loop for the page in case the eviction fails with an EBUSY error.

Reply:
Updated in bd29a083241316a3683c148705347ee90d671797